lightning/src/routing/router.rs

```diff
@@ -333,84 +592,262 @@ pub fn get_route<L: Deref>(our_node_id: &PublicKey, network: &NetworkGraph, targ
-					// flags for the last hop node itself).
-					assert!(res.last().unwrap().pubkey == *target);
+
+					if ordered_hops.last().unwrap().route_hop.pubkey == *payee {
+						break 'path_walk;
+					}
+
+					new_entry = match dist.remove(&ordered_hops.last().unwrap().route_hop.pubkey) {
+						Some(payment_hop) => payment_hop,
+						None => {
+							// If we can't reach a given node, something went wrong
+							// during path traverse.
+							// Stop looking for more paths here because next iterations
+							// will face the exact same issue.
+							break 'paths_collection;
+						}
+					};
+					// We "propagate" the fees one hop backward (topologically) here,
+					// so that fees paid on the current channels are associated with
+					// the previous channel (where they will be paid).
+					ordered_hops.last_mut().unwrap().route_hop.fee_msat = new_entry.hop_use_fee_msat;
+					ordered_hops.last_mut().unwrap().route_hop.cltv_expiry_delta = new_entry.route_hop.cltv_expiry_delta;
+					ordered_hops.push(new_entry.clone());
 				}
-				if res.last().unwrap().pubkey == *target {
-					break;
+				ordered_hops.last_mut().unwrap().route_hop.fee_msat = value_contribution_msat;
+				ordered_hops.last_mut().unwrap().route_hop.cltv_expiry_delta = final_cltv;
+
+				let payment_path = PaymentPath {hops: ordered_hops};
+				// Since a path allows to transfer as much value as
+				// the smallest channel it has ("bottleneck"), we should recompute
+				// the fees so sender HTLC don't overpay fees when traversing
+				// larger channels than the bottleneck. This may happen because
+				// when we were selecting those channels we were not aware how much value
+				// this path will transfer, and the relative fee for them
+				// might have been computed considering a larger value.
+				// Remember that we used these channels so that we don't rely
+				// on the same liquidity in future paths.
+				for (i, payment_hop) in payment_path.hops.iter().enumerate() {
+					let channel_liquidity_available_msat = bookkeeped_channels_liquidity_available_msat.get_mut(&payment_hop.route_hop.short_channel_id).unwrap();
+					let mut spent_on_hop_msat = value_contribution_msat;
+					if i != payment_path.hops.len() - 1 {
+						// For the last hop the fee is 0, so we don't have to add it here.
+						spent_on_hop_msat += payment_hop.get_fee_paid_msat();
+					}
+					if *channel_liquidity_available_msat < spent_on_hop_msat {
+						// This should not happen because we do recompute fees right before,
+						// trying to avoid cases when a hop is not usable due to the fee situation.
+						break 'path_construction;
+					}
+					*channel_liquidity_available_msat -= spent_on_hop_msat;
 				}
+				// Track the total amount all our collected paths allow to send so that we:
+				// - know when to stop looking for more paths
+				// - know which of the hops are useless considering how much more sats we need
```

Comment:
Add a comment redirection "know which of the hops are useless considering how much more sats we need (`contributes_sufficient_value`)"